File: python/src/racp/session.py
"""Transport-independent rACP v1 session semantics."""

from __future__ import annotations

import re
import time
from collections import OrderedDict
from collections.abc import Callable
from dataclasses import dataclass
from enum import Enum

from .protocol import (
    MAX_LINE_BYTES,
    Ack,
    Bye,
    Command,
    Error,
    Message,
    Ping,
    Pong,
    ProtocolError,
    State,
    Subscribe,
    Unsubscribe,
    encode_message,
    parse_message,
)

PEER_RE = re.compile(r"[A-Za-z0-9][A-Za-z0-9._-]{0,63}\Z")
CAP_RE = re.compile(r"[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*\Z")


@dataclass(frozen=True)
class Hello:
    peer_type: str
    peer_id: str
    capabilities: tuple[str, ...] = ()

    def lines(self) -> list[str]:
        _validate_peer(self.peer_type)
        _validate_peer(self.peer_id)
        caps = tuple(sorted(self.capabilities))
        if caps != self.capabilities or len(set(caps)) != len(caps):
            raise ValueError("capabilities must be sorted and unique")
        for cap in caps:
            _validate_cap(cap)
        return ["RACP/1 HELLO", f"PEER {self.peer_type} {self.peer_id}", *(f"CAP {cap}" for cap in caps), "END"]


class SessionState(Enum):
    HELLO = "hello"
    ESTABLISHED = "established"
    CLOSING = "closing"
    CLOSED = "closed"


class SessionClosed(RuntimeError):
    pass


CommandHandler = Callable[[Command], str | None]


def _validate_peer(value: str) -> None:
    if not PEER_RE.fullmatch(value):
        raise ValueError("invalid peer token")


def _validate_cap(value: str) -> None:
    if len(value) > 128 or not CAP_RE.fullmatch(value):
        raise ValueError("invalid capability")


class Session:
    def __init__(
        self,
        local: Hello,
        handler: CommandHandler | None = None,
        *,
        ledger_size: int = 1024,
        now: Callable[[], float] = time.monotonic,
    ) -> None:
        if ledger_size < 1:
            raise ValueError("ledger_size must be positive")
        local.lines()
        self.local = local
        self.peer: Hello | None = None
        self.state = SessionState.HELLO
        self.handler = handler or (lambda _command: None)
        self.ledger_size = ledger_size
        self._ledger: OrderedDict[int, tuple[Command, Ack | Error]] = OrderedDict()
        self._hello_lines: list[str] = []
        self.subscriptions: set[str] = set()
        self.state_revisions: dict[str, int] = {}
        self.malformed_count = 0
        self.now = now
        self.last_received = now()
        self.outstanding_ping: tuple[int, float] | None = None

    def hello_bytes(self) -> bytes:
        return ("\n".join(self.local.lines()) + "\n").encode()

    def receive_line(self, line: str) -> list[Message]:
        if self.state in (SessionState.CLOSING, SessionState.CLOSED):
            raise SessionClosed(self.state.value)
        self.last_received = self.now()
        if self.state is SessionState.HELLO:
            self._hello_lines.append(line)
            if len(self._hello_lines) > 1027:
                raise ProtocolError("malformed_message", fatal=True)
            if line == "END":
                self.peer = _parse_hello(self._hello_lines)
                self._hello_lines.clear()
                self.state = SessionState.ESTABLISHED
            return []
        try:
            message = parse_message(line)
        except ProtocolError as exc:
            self.malformed_count += 1
            if self.malformed_count >= 3:
                exc.fatal = True
            raise
        return self.receive(message)

    def receive(self, message: Message) -> list[Message]:
        if self.state is not SessionState.ESTABLISHED:
            raise ProtocolError("handshake_required", fatal=True)
        if isinstance(message, Ping):
            return [Pong(message.nonce)]
        if isinstance(message, Pong):
            if self.outstanding_ping and self.outstanding_ping[0] == message.nonce:
                self.outstanding_ping = None
            return []
        if isinstance(message, Bye):
            self.state = SessionState.CLOSING
            return []
        if isinstance(message, Command):
            return [self._command(message)]
        if isinstance(message, Subscribe):
            return [self._subscribe(message)]
        if isinstance(message, Unsubscribe):
            self.subscriptions.discard(message.name)
            return [Ack(message.request_id)]
        if isinstance(message, State):
            previous = self.state_revisions.get(message.name, -1)
            if message.revision > previous:
                self.state_revisions[message.name] = message.revision
            return []
        return []

    def _command(self, command: Command) -> Ack | Error:
        previous = self._ledger.get(command.request_id)
        if previous is not None:
            return previous[1] if previous[0] == command else Error(command.request_id, "request_id_conflict")
        assert self.peer is not None
        if command.name not in self.local.capabilities:
            response: Ack | Error = Error(command.request_id, "unsupported_capability")
        else:
            try:
                code = self.handler(command)
                if code:
                    try:
                        _validate_cap(code)
                    except ValueError:
                        code = "application_error"
                response = Error(command.request_id, code) if code else Ack(command.request_id)
            except Exception:
                response = Error(command.request_id, "application_error")
        self._ledger[command.request_id] = (command, response)
        while len(self._ledger) > self.ledger_size:
            self._ledger.popitem(last=False)
        return response

    def _subscribe(self, message: Subscribe) -> Ack | Error:
        if "state.subscribe" not in self.local.capabilities or message.name not in self.local.capabilities:
            return Error(message.request_id, "unsupported_capability")
        self.subscriptions.add(message.name)
        return Ack(message.request_id)

    def heartbeat(self, nonce: int) -> list[Message]:
        current = self.now()
        if self.outstanding_ping is not None:
            if current - self.outstanding_ping[1] >= 5:
                self.state = SessionState.CLOSED
                raise SessionClosed("heartbeat_timeout")
            return []
        if current - self.last_received >= 10:
            self.outstanding_ping = (nonce, current)
            return [Ping(nonce)]
        return []

    @staticmethod
    def encode(messages: list[Message]) -> bytes:
        output = bytearray()
        for message in messages:
            line = encode_message(message).encode("utf-8")
            if len(line) > MAX_LINE_BYTES:
                raise ProtocolError("line_too_long", fatal=True)
            output.extend(line)
            output.append(0x0A)
        return bytes(output)


def _parse_hello(lines: list[str]) -> Hello:
    if len(lines) < 3 or lines[0] != "RACP/1 HELLO" or lines[-1] != "END":
        code = "unsupported_version" if lines and lines[0].startswith("RACP/") else "malformed_message"
        raise ProtocolError(code, fatal=True)
    peer_parts = lines[1].split(" ")
    if len(peer_parts) != 3 or peer_parts[0] != "PEER":
        raise ProtocolError("malformed_message", fatal=True)
    try:
        _validate_peer(peer_parts[1])
        _validate_peer(peer_parts[2])
    except ValueError as exc:
        raise ProtocolError("malformed_message", fatal=True) from exc
    caps: list[str] = []
    for line in lines[2:-1]:
        parts = line.split(" ")
        if len(parts) != 2 or parts[0] != "CAP":
            raise ProtocolError("malformed_message", fatal=True)
        try:
            _validate_cap(parts[1])
        except ValueError as exc:
            raise ProtocolError("malformed_message", fatal=True) from exc
        caps.append(parts[1])
    if caps != sorted(set(caps)):
        raise ProtocolError("malformed_message", fatal=True)
    return Hello(peer_parts[1], peer_parts[2], tuple(caps))

File: Sources/ReasonableACP/Session.swift
import Foundation

public struct RACPHello: Sendable, Equatable {
  public let peerType: String
  public let peerID: String
  public let capabilities: [String]

  public init(peerType: String, peerID: String, capabilities: [String] = []) throws {
    try validatePeer(peerType)
    try validatePeer(peerID)
    guard capabilities == capabilities.sorted(), Set(capabilities).count == capabilities.count
    else {
      throw RACPProtocolError.malformedMessage()
    }
    for capability in capabilities { try validateName(capability) }
    self.peerType = peerType
    self.peerID = peerID
    self.capabilities = capabilities
  }

  public var lines: [String] {
    ["RACP/1 HELLO", "PEER \(peerType) \(peerID)"] + capabilities.map { "CAP \($0)" } + ["END"]
  }

  public var encoded: Data { Data((lines.joined(separator: "\n") + "\n").utf8) }
}

public enum RACPSessionState: Sendable, Equatable {
  case hello
  case established
  case closing
  case closed
}

public enum RACPSessionError: Error, Sendable, Equatable {
  case closed(String)
  case asynchronousCommandHandler
}

public enum RACPCommandDisposition: Sendable, Equatable {
  case success
  case error(String)
}

public enum RACPSubscriptionEvent: Sendable, Equatable {
  case subscribed(String)
  case unsubscribed(String)
}

public final class RACPSession {
  public typealias CommandHandler = (Command) -> String?
  public typealias AsyncCommandHandler = (Command) async throws -> RACPCommandDisposition
  public typealias StateHandler = (StateMessage) -> Void

  public let local: RACPHello
  public private(set) var peer: RACPHello?
  public private(set) var state: RACPSessionState = .hello
  public private(set) var subscriptions: Set<String> = []
  public private(set) var stateRevisions: [String: UInt64] = [:]
  public private(set) var malformedCount = 0
  public private(set) var lastReceived: TimeInterval
  public private(set) var outstandingPing: (nonce: UInt64, sentAt: TimeInterval)?
  public private(set) var lastMessage: RACPMessage?
  public private(set) var lastSubscriptionEvent: RACPSubscriptionEvent?

  private let handler: CommandHandler?
  private let asyncHandler: AsyncCommandHandler?
  private let stateHandler: StateHandler
  private let ledgerSize: Int
  private let now: () -> TimeInterval
  private var helloLines: [String] = []
  private var ledger: [UInt64: LedgerEntry] = [:]
  private var ledgerOrder: [UInt64] = []

  private struct LedgerEntry {
    let command: Command
    let response: RACPMessage
  }

  public init(
    local: RACPHello,
    ledgerSize: Int = 1_024,
    now: @escaping () -> TimeInterval = { ProcessInfo.processInfo.systemUptime },
    commandHandler: @escaping CommandHandler = { _ in nil },
    stateHandler: @escaping StateHandler = { _ in }
  ) {
    precondition(ledgerSize > 0)
    self.local = local
    self.ledgerSize = ledgerSize
    self.now = now
    handler = commandHandler
    asyncHandler = nil
    self.stateHandler = stateHandler
    lastReceived = now()
  }

  public init(
    local: RACPHello,
    ledgerSize: Int = 1_024,
    now: @escaping () -> TimeInterval = { ProcessInfo.processInfo.systemUptime },
    asyncCommandHandler: @escaping AsyncCommandHandler,
    stateHandler: @escaping StateHandler = { _ in }
  ) {
    precondition(ledgerSize > 0)
    self.local = local
    self.ledgerSize = ledgerSize
    self.now = now
    handler = nil
    asyncHandler = asyncCommandHandler
    self.stateHandler = stateHandler
    lastReceived = now()
  }

  public func receive(line: String) throws -> [RACPMessage] {
    guard state != .closing, state != .closed else { throw RACPSessionError.closed("closed") }
    lastReceived = now()
    lastSubscriptionEvent = nil
    if state == .hello {
      lastMessage = nil
      helloLines.append(line)
      guard helloLines.count <= 1_027 else { throw RACPProtocolError.malformedMessage(fatal: true) }
      if line == "END" {
        peer = try Self.parseHello(helloLines)
        helloLines.removeAll(keepingCapacity: false)
        state = .established
      }
      return []
    }
    do {
      let message = try RACPMessage.parse(line)
      lastMessage = message
      return try receive(message: message)
    } catch let error as RACPProtocolError {
      malformedCount += 1
      if malformedCount >= 3 { throw RACPProtocolError.malformedMessage(fatal: true) }
      throw error
    }
  }

  /// Receives one wire line, awaiting application command completion before
  /// returning its terminal ACK or ERR response.
  public nonisolated(nonsending) func receiveAsync(line: String) async throws -> [RACPMessage] {
    guard state != .closing, state != .closed else { throw RACPSessionError.closed("closed") }
    lastReceived = now()
    lastSubscriptionEvent = nil
    if state == .hello {
      return try receiveHello(line: line)
    }
    do {
      let message = try RACPMessage.parse(line)
      lastMessage = message
      return try await receiveAsync(message: message)
    } catch let error as RACPProtocolError {
      malformedCount += 1
      if malformedCount >= 3 { throw RACPProtocolError.malformedMessage(fatal: true) }
      throw error
    }
  }

  public func receive(message: RACPMessage) throws -> [RACPMessage] {
    guard state == .established else { throw RACPProtocolError.handshakeRequired }
    lastSubscriptionEvent = nil
    switch message {
    case .ping(let nonce): return [.pong(nonce)]
    case .pong(let nonce):
      if outstandingPing?.nonce == nonce { outstandingPing = nil }
      return []
    case .bye:
      state = .closing
      return []
    case .command(let command):
      guard asyncHandler == nil else { throw RACPSessionError.asynchronousCommandHandler }
      return [handle(command)]
    case .subscribe(let id, let name): return [subscribe(id: id, name: name)]
    case .unsubscribe(let id, let name):
      subscriptions.remove(name)
      lastSubscriptionEvent = .unsubscribed(name)
      return [.ack(id)]
    case .state(let message):
      let previous = stateRevisions[message.name]
      if previous == nil || message.revision > previous! {
        stateRevisions[message.name] = message.revision
        stateHandler(message)
      }
      return []
    case .ack, .error: return []
    }
  }

  public nonisolated(nonsending) func receiveAsync(message: RACPMessage) async throws
    -> [RACPMessage]
  {
    guard state == .established else { throw RACPProtocolError.handshakeRequired }
    lastSubscriptionEvent = nil
    if case .command(let command) = message { return [await handleAsync(command)] }
    return try receive(message: message)
  }

  public func heartbeat(nonce: UInt64) throws -> [RACPMessage] {
    let current = now()
    if let outstandingPing {
      if current - outstandingPing.sentAt >= 5 {
        state = .closed
        throw RACPSessionError.closed("heartbeat_timeout")
      }
      return []
    }
    if current - lastReceived >= 10 {
      try validateRequestIDForSession(nonce)
      outstandingPing = (nonce, current)
      return [.ping(nonce)]
    }
    return []
  }

  public func markClosed() { state = .closed }

  public static func encode(_ messages: [RACPMessage]) throws -> Data {
    var output = Data()
    for message in messages {
      let line = try message.encoded()
      guard line.utf8.count <= racpMaximumLineBytes else { throw RACPProtocolError.lineTooLong }
      output.append(contentsOf: line.utf8)
      output.append(0x0A)
    }
    return output
  }

  private func handle(_ command: Command) -> RACPMessage {
    if let previous = ledger[command.requestID] {
      return previous.command == command
        ? previous.response : .error(command.requestID, "request_id_conflict")
    }
    let response: RACPMessage
    if !local.capabilities.contains(command.name) {
      response = .error(command.requestID, "unsupported_capability")
    } else if let code = handler?(command) {
      response = validatedError(id: command.requestID, code: code)
    } else {
      response = .ack(command.requestID)
    }
    record(command: command, response: response)
    return response
  }

  private nonisolated(nonsending) func handleAsync(_ command: Command) async -> RACPMessage {
    if let previous = ledger[command.requestID] {
      return previous.command == command
        ? previous.response : .error(command.requestID, "request_id_conflict")
    }
    let response: RACPMessage
    if !local.capabilities.contains(command.name) {
      response = .error(command.requestID, "unsupported_capability")
    } else if let asyncHandler {
      do {
        switch try await asyncHandler(command) {
        case .success: response = .ack(command.requestID)
        case .error(let code): response = validatedError(id: command.requestID, code: code)
        }
      } catch {
        response = .error(command.requestID, "application_error")
      }
    } else if let code = handler?(command) {
      response = validatedError(id: command.requestID, code: code)
    } else {
      response = .ack(command.requestID)
    }
    record(command: command, response: response)
    return response
  }

  private func validatedError(id: UInt64, code: String) -> RACPMessage {
    do {
      try validateName(code)
      return .error(id, code)
    } catch {
      return .error(id, "application_error")
    }
  }

  private func record(command: Command, response: RACPMessage) {
    ledger[command.requestID] = LedgerEntry(command: command, response: response)
    ledgerOrder.append(command.requestID)
    if ledgerOrder.count > ledgerSize { ledger.removeValue(forKey: ledgerOrder.removeFirst()) }
  }

  private func subscribe(id: UInt64, name: String) -> RACPMessage {
    guard local.capabilities.contains("state.subscribe"), local.capabilities.contains(name) else {
      return .error(id, "unsupported_capability")
    }
    subscriptions.insert(name)
    lastSubscriptionEvent = .subscribed(name)
    return .ack(id)
  }

  private func receiveHello(line: String) throws -> [RACPMessage] {
    lastMessage = nil
    helloLines.append(line)
    guard helloLines.count <= 1_027 else { throw RACPProtocolError.malformedMessage(fatal: true) }
    if line == "END" {
      peer = try Self.parseHello(helloLines)
      helloLines.removeAll(keepingCapacity: false)
      state = .established
    }
    return []
  }

  private static func parseHello(_ lines: [String]) throws -> RACPHello {
    guard lines.count >= 3, lines.last == "END" else {
      throw RACPProtocolError.malformedMessage(fatal: true)
    }
    guard lines[0] == "RACP/1 HELLO" else {
      if lines[0].hasPrefix("RACP/") { throw RACPProtocolError.unsupportedVersion }
      throw RACPProtocolError.malformedMessage(fatal: true)
    }
    let peer = lines[1].split(separator: " ", omittingEmptySubsequences: false).map(String.init)
    guard peer.count == 3, peer[0] == "PEER" else {
      throw RACPProtocolError.malformedMessage(fatal: true)
    }
    var capabilities: [String] = []
    for line in lines.dropFirst(2).dropLast() {
      let fields = line.split(separator: " ", omittingEmptySubsequences: false).map(String.init)
      guard fields.count == 2, fields[0] == "CAP" else {
        throw RACPProtocolError.malformedMessage(fatal: true)
      }
      capabilities.append(fields[1])
    }
    do {
      return try RACPHello(peerType: peer[1], peerID: peer[2], capabilities: capabilities)
    } catch {
      throw RACPProtocolError.malformedMessage(fatal: true)
    }
  }
}

private func validateRequestIDForSession(_ value: UInt64) throws {
  guard value > 0, value <= racpMaximumSafeInteger else {
    throw RACPProtocolError.malformedMessage()
  }
}
